provisioning/image_cache.go

```diff
@@ -60,27 +61,55 @@ func imageVolume() corev1.Volume {
 	}
 }
 
-func imageCacheConfig(targetNamespace string, config metal3iov1alpha1.ProvisioningSpec) (*metal3iov1alpha1.ProvisioningSpec, error) {
-	// The download URL looks something like:
-	// https://releases-art-rhcos.svc.ci.openshift.org/art/storage/releases/rhcos-4.2/42.80.20190725.1/rhcos-42.80.20190725.1-openstack.qcow2?sha256sum=123
-	downloadURL, err := url.Parse(config.ProvisioningOSDownloadURL)
+// Helper to transform the first level (metal3 pod) cache URLs to second level
+// (control-plane daemonset) cache
+func transformURL(targetNamespace, URL string) (string, error) {
+	downloadURL, err := url.Parse(URL)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
 	imageName := path.Base(fileCompressionSuffix.ReplaceAllString(downloadURL.Path, ""))
 
-	// The first level of cache downloads the file and makes an uncompressed
-	// version of it available to this second-level cache somewhere like:
-	// http://metal3-state.openshift-machine-api:6180/images/rhcos-42.80.20190725.1-openstack.qcow2/rhcos-42.80.20190725.1-openstack.qcow2
+	// The first level of cache downloads the file and makes an version of it available
+	// to this second-level cache which then transforms it into the following format:
+	// Original format: http://<URL>:6180/images/rhcos-42.80.20190725.1-openstack.qcow2/cached-rhcos-42.80.20190725.1-openstack.qcow2
+	// Transformed format: http://metal3-state.openshift-machine-api:6181/images/rhcos-42.80.20190725.1-openstack.qcow2/cached-rhcos-42.80.20190725.1-openstack.qcow2
```

Comment:
I don't believe either of these is correct. The _original_ format (i.e. the URL that is being passed in to this function) is something like:

```
https://releases-art-rhcos.svc.ci.openshift.org/art/storage/releases/rhcos-4.2/42.80.20190725.1/rhcos-42.80.20190725.1-openstack.qcow2.gz?sha256sum=123
```

We then transform it into:

```
http://metal3-state.openshift-machine-api.svc.cluster.local:6180/images/rhcos-42.80.20190725.1-openstack.qcow2/cached-rhcos-42.80.20190725.1-openstack.qcow2
```

which is where the first-level cache serves it from. Finally, the second-level cache will make it available at:

```
http://cluster.local:6181/images/rhcos-42.80.20190725.1-openstack.qcow2/cached-rhcos-42.80.20190725.1-openstack.qcow2
```

The previous comments were considerably more accurate, although neither was entirely correct.

Reply:
The comment makes use of the word 'Original', which is unfortunate not to mention inaccurate. I should have mentioned first level cache url instead of the original. I'll rewrite the comment.

Reply:
Thanks for clarifying the comments. I've updated them.